analysers/analyser_osmosis_powerline.py

```diff
@@ -24,28 +24,110 @@
 from modules.Stablehash import stablehash64
 
-# Power lines nodes with their voltage as array padded up to 99 zeros
+# Power lines nodes with their voltage as array padded up to 99 zeros to cope with non-numerical values
 # Lines with no voltage get null voltage instead empty array
 sql01 = """
 CREATE TEMP TABLE power_lines_nodes AS
 SELECT
-    ways.id as wid,
-    unnest(ways.nodes) AS nid,
-    unnest(ways.nodes[2:array_length(ways.nodes,1)]) AS nid_next,
-    ways.tags->'cables' as cables,
-    ways.tags->'circuits' as circuits,
-    coalesce(ways.tags->'location','overhead') as location,
+    w.id as wid,
+    unnest('{NULL}' || w.nodes[1:array_length(w.nodes, 1) - 1]) AS nid_prec,
+    unnest(w.nodes) AS nid,
+    unnest(w.nodes[2:array_length(w.nodes, 1)]) AS nid_next,
+    w.tags->'cables' AS cables,
+    coalesce((w.tags->'circuits')::integer, 1) AS circuits,
+    coalesce(w.tags->'location', 'overhead') AS location,
     voltage
 FROM
-    ways
-    JOIN LATERAL (SELECT array_agg(lpad(v, 99, '0')) FROM unnest(regexp_split_to_array(ways.tags->'voltage','; *')) AS t(v)) AS t(voltage) ON TRUE
+    ways AS w
+    JOIN LATERAL (
+        SELECT array_agg(lpad(v, 99, '0'))
+        FROM unnest(array_cat(
+            array_fill(
+                substring(w.tags->'voltage' for greatest(char_length (w.tags->'voltage'), position(';' in w.tags->'voltage')))::text, -- voltage1 in voltage1;voltage2
+                ARRAY[coalesce((w.tags->'circuits')::integer, 1) - 1 + char_length(coalesce(w.tags->'voltage', '')) - char_length(replace(coalesce(w.tags->'voltage',''), ';', ''))]
+            ),
+            regexp_split_to_array(w.tags->'voltage', '; *'))
+        ) AS t(v)) AS t(voltage)
+        ON TRUE
 WHERE
-    ways.tags != ''::hstore AND
-    ways.tags?'power' AND
-    ways.tags->'power' IN ('line', 'minor_line', 'cable')
+    w.tags != ''::hstore AND
+    w.tags?'power' AND
+    w.tags->'power' IN ('line', 'minor_line', 'cable') AND
+    w.tags->'voltage' IS NOT NULL
+
+UNION ALL
+
+SELECT
+    w.id AS wid,
+    unnest('{NULL}' || w.nodes[1:array_length(w.nodes, 1) - 1]) AS nid_prec,
+    unnest(w.nodes) AS nid,
+    unnest(w.nodes[2:array_length(w.nodes, 1)]) AS nid_next,
+    w.tags->'cables' AS cables,
+    coalesce((w.tags->'circuits')::integer, 1) AS circuits,
+    coalesce(w.tags->'location', 'overhead') AS location,
+    NULL AS voltage
+FROM
+   ways AS w
+WHERE
+    w.tags != ''::hstore AND
+    w.tags?'power' AND
+    w.tags->'power' IN ('line', 'minor_line', 'cable') AND
+    w.tags->'voltage' IS NULL
+"""
+
+# Build junctions knowledge
+# Topoedges are couples attached to a given node with their neighbors.
+# Topoedges are agregated by nodes and location (two *overhead* lines between two node will give a single topoedge)
+# Involved nodes are not necessary power, particularly on cables
+sql02 = """
+CREATE TEMP TABLE power_lines_topoedges AS
+WITH topotuples as (
+    SELECT
+        n.wid,
+        n.nid # n.nid_prec AS tid,
```

Comment:
`#` is really an intersection operator here ?

Reply:
It's a XOR.
We only need diversity at nid scale, not the whole space.
And there is no commutative operation that doesn't overflow bigint limit (bigint + bigint or bigint * bigint aren't fine)